### ISSUE TITLE
Hybrid Neumann and Dirichlet type boundary conditions.

### ISSUE THREAD

**Issue by shaoyaoqian** [gauvain-thomas/Geometric-Multigrid#1]
I think it is not practical to use two kinds of grids, but I can not find a way to solve this problem. Any sugestions?

**Comment by gauvain-thomas** [comment]
Only relaxation methods for vertex-centered grid with Dirichlet boundary conditions and cell-centered grid with Neumann boundary conditions are implemented to this day. To add mixed boundary conditions, you would have to implement specific smoothing and residual operators.

I believe the easiest way to do this would be on a cell-centered grid would be to change the padding mode on sides with Dirichlet boundary conditions, setting the outer edge to the opposite of the inner edge, thus forcing the middle point to be zero (homogeneous conditions). This is supposing keeping Dirichlet conditions on vertices and Neumann conditions cell-centered.

Hope this answers your question.

**Comment by shaoyaoqian** [comment]
Thanks!

I just read a book called "A Multigrid Tutorial", I saw this figure:
<img width="490" alt="image" src="https://user-images.githubusercontent.com/115222128/227978347-957f22aa-976e-476d-9370-0178c62644a8.png">
I noticed these points:
1. Restricttion and Prolongation: All boundaries are set as 'replicate' padding mode. 
2. Smoother: Relaxation are performed on interior points and Neumann points and Dirichlet points are ignored.

**Comment by gauvain-thomas** [comment]
The 'replicate' padding mode asserts a zero derivative condition for a cell-centered grid, as the discretized derivative lies between the two points (so on the edge). For Dirichlet condition, the edges are fixed to zero, and this is done by only ever working on the inside of the grid.

I read that book too while making this code, and I also recommend "Multigrid" by Ulrich Trottenberg for further explanations.

**Comment by shaoyaoqian** [comment]
Thank you again for your recommendation! I have successfully implemented hybrid boundary conditions after reading the book you suggested. I found that using vertex-centered grids made the implementation process easier, as both FW and HW could be utilized. Additionally, aligning the Neumann boundary with the physical boundary was also made possible.

<img width="826" alt="image" src="https://user-images.githubusercontent.com/115222128/228236404-c3c0c4ee-12f6-458b-9b77-e59acefe63ff.png">

Thanks again!
